### PR TITLE
Implements Best Match Locator

### PR DESCRIPTION
**Pull request by mahalay**
* Replaces PrioritizedResourceManager
* Adds `suite` option to describe command
* Potentially renders ResourceLocator::getPriority() deprecated

============================================================
Composer allows for classes of the same namespace to be located on multiple paths. Phpspec makes the decision which locator to use based on `priority`. It would be better if the developer is informed that a class being described matches multiple suites and thus be allowed to specify a suite to use. This will make Phpspec easier to work with when used on projects using old frameworks (e.g., Code Igniter).

### CODE REVIEW

File: src/PhpSpec/Locator/BestMatchResourceManager.php [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    /**
+     * @param  string $classname
+     *
+     * @return array
+     */
+    private function sortAndGroupLocatorsByMatchScore($classname)
+    {
+        $matchedLocators = [];
+        foreach ($this->locators as $locator) {
+            if (($matchScore = $matchScore = $locator->calculateMatchScore($classname)) > 0) {
```
Comment (mheki) [inline src/PhpSpec/Locator/BestMatchResourceManager.php]: `$matchScore = $matchScore = ` mistake here

File: src/PhpSpec/Locator/BestMatchResourceManager.php [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    private $locators = array();
+
+    /**
+     * @param ResourceLocator $locator
+     */
+    public function registerLocator(ResourceLocator $locator)
+    {
+        $this->locators[] = $locator;
+
+        @usort($this->locators, function (ResourceLocator $locator1, ResourceLocator $locator2) {
```
Comment (mheki) [inline src/PhpSpec/Locator/BestMatchResourceManager.php]: out of curiosity - why `@` here?

File: spec/PhpSpec/Locator/BestMatchResourceManagerSpec.php [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+        $resource1->getSpecClassname()->willReturn('Some\Spec1');
+        $resource2->getSpecClassname()->willReturn('Some\Spec2');
+        $resource3->getSpecClassname()->willReturn('Some\Spec3');
+
+        $this->locateResources('s:query')->shouldReturn(array($resource1, $resource3, $resource2));
+    }
+
+    function it_locates_all_locators_resources_if_query_string_is_empty($locator1, $locator2,
+                                                                        Resource $resource1, Resource $resource2, Resource $resource3
```
Comment (mheki) [inline spec/PhpSpec/Locator/BestMatchResourceManagerSpec.php]: something is messed up with whitespaces here, could you make it more PSR-2 compliant, please?

File: src/PhpSpec/Locator/BestMatchResourceManager.php (hunk 2 above)
Comment (ciaranmcnulty) [inline src/PhpSpec/Locator/BestMatchResourceManager.php]: There's some historical issue that needs it - I think PHP believes the array has changed while being sorted and issues a warning?

**Comment by mahalay** [comment]
Thanks for the feedback. Corrections were made as suggested.

**Comment by mheki** [comment]
thanks @mahalay! Would be great if new features were in sync with the docs. Can you describe your changes in docs, please?

**Comment by mahalay** [comment]
Will work on it after office... Why have I forgotten to update the docs. ;)

**Comment by mahalay** [comment]
hey @mheki, I've uncovered a scenario where the current implementation will fail. I'll fix that first then finish the docs.

**Comment by mahalay** [comment]
Updated the docs and added the scenario where the implementation was failing.

Just curious, what do you guys use for coding standard testing? I tried using squizlabs/php_codesniffer and specified PSR2 as the coding standard and there are a lot of codes with warnings.

**Comment by ciaranmcnulty** [comment]
We're using the scrutiniser PRS-2 rules

**Comment by ciaranmcnulty** [comment]
@mahalay can you resolve the conflicts please?

**Comment by ciaranmcnulty** [comment]
Something now seems to be breaking on PHP 7.2

**Comment by mahalay** [comment]
Hey @ciaranmcnulty, installed PHP 7.2 from my end but did not get any errors when I run the tests. Will try to figure out the issue.

**Comment by ciaranmcnulty** [comment]
Ah it looks like it's the dev dependencies build that's failing https://travis-ci.org/phpspec/phpspec/jobs/288129896

**Comment by mahalay** [comment]
The error was due to the Exception class not being displayed along with the exception message.

**Comment by ciaranmcnulty** [comment]
Can you rebase this please?

**Comment by mahalay** [comment]
Rebase done, but fails tests on Windows - will setup up a Win VM later

**Pull request closed without merging**